flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -958,10 +960,10 @@ private void configureCheckpointing() {
 				.setMinPauseBetweenCheckpoints(cfg.getMinPauseBetweenCheckpoints())
 				.setMaxConcurrentCheckpoints(cfg.getMaxConcurrentCheckpoints())
 				.setCheckpointRetentionPolicy(retentionAfterTermination)
-				.setExactlyOnce(getCheckpointingMode(cfg) == CheckpointingMode.EXACTLY_ONCE)
+				.setExactlyOnce(exactlyOnce)
 				.setPreferCheckpointForRecovery(cfg.isPreferCheckpointForRecovery())
 				.setTolerableCheckpointFailureNumber(cfg.getTolerableCheckpointFailureNumber())
-				.setUnalignedCheckpointsEnabled(cfg.isUnalignedCheckpointsEnabled())
+				.setUnalignedCheckpointsEnabled(exactlyOnce && cfg.isUnalignedCheckpointsEnabled())
```

Comment:
This check seems duplicated `exactlyOnce && cfg.isUnalignedCheckpointsEnabled()` in two places. Maybe hide `exactlyOnce &&` check inside `isUnalignedCheckpointsEnabled`?

Reply:
Also possible way, it's pretty much ignoring UC when at least once is set. See also above for a deeper discussion.